packages/kit/src/runtime/server/endpoint.js

```diff
@@ -51,16 +54,30 @@ export async function render_endpoint(event, mod, state) {
 		}
 
 		return response;
-	} catch (error) {
-		if (error instanceof HttpError) {
-			return new Response(error.message, { status: error.status });
-		} else if (error instanceof Redirect) {
+	} catch (e) {
+		const error = normalize_error(e);
+
+		if (error instanceof Redirect) {
 			return new Response(undefined, {
 				status: error.status,
-				headers: { Location: error.location }
+				headers: { location: error.location }
 			});
-		} else {
-			throw error;
 		}
+
+		if (!(error instanceof HttpError)) {
+			options.handle_error(error, event);
+		}
+
+		const wants_json =
+			negotiate(event.request.headers.get('accept') || 'text/html', [
+				'text/html',
+				'application/json'
+			]) === 'application/json';
+
+		return wants_json
+			? json(error_to_pojo(error, options.get_stack), {
+					status: error instanceof HttpError ? error.status : 500
+			  })
+			: new Response(error.message, { status: error instanceof HttpError ? error.status : 500 });
```

Comment:
Is there a reason not to render the HTML error page here? Feels like if request doesn't specifically ask for JSON we should just do that.

An alternative could be to render the HTML response if `sec-fetch-dest` is `document`, otherwise JSON

Reply:
Mhm true we can reuse the static error page here. I didn't know about `sec-fetch-dest`, I did what was done in a different but related code path where there's also a check to "is this a JSON request".